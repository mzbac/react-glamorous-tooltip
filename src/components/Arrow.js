import glamorous from 'glamorous';
export default glamorous.span(props => {
  const { fg, bg, position, arrow, arrowStyle, transitionTime } = props;
  const baseStyle = {
    position: 'absolute',
    content: '""',
    transition: `all ${transitionTime} ease-in-out`,
  };
  let fgStyle = {};
  let bgStyle = {};
  const newArrowStyle = { color: '#fff', borderColor: 'rgba(0,0,0,.4)', ...arrowStyle };
  if (fg) {
    fgStyle.zIndex = 60;
    const fgColorBorder = `10px solid ${newArrowStyle.color}`;
    const fgTransBorder = '8px solid transparent';
    const fgStyleLeftOrRightBase = {
      borderTop: fgTransBorder,
      borderBottom: fgTransBorder,
      top: 'calc(50% - 8px)',
    };
    if (position === 'left') {
      fgStyle = { ...fgStyle, ...fgStyleLeftOrRightBase, right: -10, borderLeft: fgColorBorder };
    }
    if (position === 'right') {
      fgStyle = { ...fgStyle, ...fgStyleLeftOrRightBase, left: -10, borderRight: fgColorBorder };
    }
    if (position === 'left' || position === 'right') {
      if (arrow === 'top') {
        fgStyle.top = 2;
      }
      if (arrow === 'bottom') {
        fgStyle.top = null;
        fgStyle.bottom = 1;
      }
    }

    const fgStyleTopOrBottomBase = {
      left: '50%',
      marginLeft: -8,
      borderLeft: fgTransBorder,
      borderRight: fgTransBorder,
    };
    if (position === 'top') {
      fgStyle = { ...fgStyle, ...fgStyleTopOrBottomBase, bottom: -10, borderTop: fgColorBorder };
    }
    if (position === 'bottom') {
      fgStyle = { ...fgStyle, ...fgStyleTopOrBottomBase, top: -10, borderBottom: fgColorBorder };
    }
    if (position === 'top' || position === 'bottom') {
      if (arrow === 'right') {
        fgStyle.left = null;
        fgStyle.right = '0px';
        fgStyle.marginLeft = null;
      }
      if (arrow === 'left') {
        fgStyle.left = null;
        fgStyle.marginLeft = null;
      }
    }
    return { ...baseStyle, ...fgStyle };
  }
  if (bg) {
    bgStyle.zIndex = 55;
    const bgBorderColor = newArrowStyle.borderColor ? newArrowStyle.borderColor : 'transparent';
    const bgColorBorder = `11px solid ${bgBorderColor}`;
    const bgTransBorder = '9px solid transparent';
    const bgStyleLeftOrRightBase = {
      borderTop: bgTransBorder,
      borderBottom: bgTransBorder,
      top: 'calc(50% - 9px)',
    };
    if (position === 'left') {
      bgStyle = { ...bgStyle, ...bgStyleLeftOrRightBase, right: -11, borderLeft: bgColorBorder };
    }
    if (position === 'right') {
      bgStyle = { ...bgStyle, ...bgStyleLeftOrRightBase, left: -11, borderRight: bgColorBorder };
    }
    if (position === 'left' || position === 'right') {
      if (arrow === 'top') {
        bgStyle.top = 1;
      }
      if (arrow === 'bottom') {
        bgStyle.top = null;
        bgStyle.bottom = 0;
      }
    }
    const bgStyleTopOrBottomBase = {
      left: '50%',
      marginLeft: -9,
      borderLeft: bgTransBorder,
      borderRight: bgTransBorder,
    };
    if (position === 'top') {
      bgStyle = { ...bgStyle, ...bgStyleTopOrBottomBase, bottom: -11, borderTop: bgColorBorder };
    }
    if (position === 'bottom') {
      bgStyle = { ...bgStyle, ...bgStyleTopOrBottomBase, top: -11, borderBottom: bgColorBorder };
    }
    if (position === 'top' || position === 'bottom') {
      if (arrow === 'right') {
        bgStyle.left = null;
        bgStyle.right = -1;
        bgStyle.marginLeft = null;
      }
      if (arrow === 'left') {
        bgStyle.left = -1;
        bgStyle.marginLeft = null;
      }
    }
    return { ...baseStyle, ...bgStyle };
  }
  return null;
});