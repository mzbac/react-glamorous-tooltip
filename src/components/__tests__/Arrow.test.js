import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import serializer from 'jest-glamor-react'
import Arrow from '../Arrow';

expect.addSnapshotSerializer(serializer);

describe('Arrow', () => {
  function renderArrow(arrowType, position, arrow) {
    test(`should render ${arrowType} Arrow with position ${position} and arrow ${arrow}`, () => {
      const props = {
        position,
        arrow,
        fg: arrowType === 'fg',
        bg: arrowType === 'bg'
      };
      const tree = mount(<Arrow {...props} />);
      expect(toJson(tree)).toMatchSnapshot();
    });
  };
  test('should render fg Arrow with default props', () => {
    const tree = mount(<Arrow fg />);
    expect(toJson(tree)).toMatchSnapshot();
  });

  renderArrow('fg','top','left');
  renderArrow('fg','top','center');
  renderArrow('fg','top','right');

  renderArrow('fg','bottom','left');
  renderArrow('fg','bottom','center');
  renderArrow('fg','bottom','right');

  renderArrow('fg','left','top');
  renderArrow('fg','left','center');
  renderArrow('fg','left','bottom');

  renderArrow('fg','right','top');
  renderArrow('fg','right','center');
  renderArrow('fg','right','bottom');

  test('should render bg Arrow with default props', () => {
    const tree = mount(<Arrow bg />);
    expect(toJson(tree)).toMatchSnapshot();
  });

  renderArrow('bg','top','left');
  renderArrow('bg','top','center');
  renderArrow('bg','top','right');

  renderArrow('bg','bottom','left');
  renderArrow('bg','bottom','center');
  renderArrow('bg','bottom','right');

  renderArrow('bg','left','top');
  renderArrow('bg','left','center');
  renderArrow('bg','left','bottom');

  renderArrow('bg','right','top');
  renderArrow('bg','right','center');
  renderArrow('bg','right','bottom');

});
