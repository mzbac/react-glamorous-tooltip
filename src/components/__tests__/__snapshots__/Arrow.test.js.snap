// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Arrow should render bg Arrow with default props 1`] = `
.glamor-0,
[data-glamor-0] {
  position: absolute;
  content: "";
  transition: all undefined ease-in-out;
  z-index: 55;
  -webkit-transition: all undefined ease-in-out;
  -moz-transition: all undefined ease-in-out;
}

<glamorous(span)
  bg={true}
>
  <span
    className="glamor-0"
  />
</glamorous(span)>
`;

exports[`Arrow should render bg Arrow with position bottom and arrow center 1`] = `
.glamor-0,
[data-glamor-0] {
  position: absolute;
  content: "";
  transition: all undefined ease-in-out;
  z-index: 55;
  left: 50%;
  margin-left: -9px;
  border-left: 9px solid transparent;
  border-right: 9px solid transparent;
  top: -11px;
  border-bottom: 11px solid rgba(0,0,0,.4);
  -webkit-transition: all undefined ease-in-out;
  -moz-transition: all undefined ease-in-out;
}

<glamorous(span)
  arrow="center"
  bg={true}
  fg={false}
  position="bottom"
>
  <span
    className="glamor-0"
  />
</glamorous(span)>
`;

exports[`Arrow should render bg Arrow with position bottom and arrow left 1`] = `
.glamor-0,
[data-glamor-0] {
  position: absolute;
  content: "";
  transition: all undefined ease-in-out;
  z-index: 55;
  left: -1px;
  border-left: 9px solid transparent;
  border-right: 9px solid transparent;
  top: -11px;
  border-bottom: 11px solid rgba(0,0,0,.4);
  -webkit-transition: all undefined ease-in-out;
  -moz-transition: all undefined ease-in-out;
}

<glamorous(span)
  arrow="left"
  bg={true}
  fg={false}
  position="bottom"
>
  <span
    className="glamor-0"
  />
</glamorous(span)>
`;

exports[`Arrow should render bg Arrow with position bottom and arrow right 1`] = `
.glamor-0,
[data-glamor-0] {
  position: absolute;
  content: "";
  transition: all undefined ease-in-out;
  z-index: 55;
  border-left: 9px solid transparent;
  border-right: 9px solid transparent;
  top: -11px;
  border-bottom: 11px solid rgba(0,0,0,.4);
  right: -1px;
  -webkit-transition: all undefined ease-in-out;
  -moz-transition: all undefined ease-in-out;
}

<glamorous(span)
  arrow="right"
  bg={true}
  fg={false}
  position="bottom"
>
  <span
    className="glamor-0"
  />
</glamorous(span)>
`;

exports[`Arrow should render bg Arrow with position left and arrow bottom 1`] = `
.glamor-0,
[data-glamor-0] {
  position: absolute;
  content: "";
  transition: all undefined ease-in-out;
  z-index: 55;
  border-top: 9px solid transparent;
  border-bottom: 9px solid transparent;
  right: -11px;
  border-left: 11px solid rgba(0,0,0,.4);
  bottom: 0;
  -webkit-transition: all undefined ease-in-out;
  -moz-transition: all undefined ease-in-out;
}

<glamorous(span)
  arrow="bottom"
  bg={true}
  fg={false}
  position="left"
>
  <span
    className="glamor-0"
  />
</glamorous(span)>
`;

exports[`Arrow should render bg Arrow with position left and arrow center 1`] = `
.glamor-0,
[data-glamor-0] {
  position: absolute;
  content: "";
  transition: all undefined ease-in-out;
  z-index: 55;
  border-top: 9px solid transparent;
  border-bottom: 9px solid transparent;
  top: calc(50% - 9px);
  right: -11px;
  border-left: 11px solid rgba(0,0,0,.4);
  -webkit-transition: all undefined ease-in-out;
  -moz-transition: all undefined ease-in-out;
}

<glamorous(span)
  arrow="center"
  bg={true}
  fg={false}
  position="left"
>
  <span
    className="glamor-0"
  />
</glamorous(span)>
`;

exports[`Arrow should render bg Arrow with position left and arrow top 1`] = `
.glamor-0,
[data-glamor-0] {
  position: absolute;
  content: "";
  transition: all undefined ease-in-out;
  z-index: 55;
  border-top: 9px solid transparent;
  border-bottom: 9px solid transparent;
  top: 1px;
  right: -11px;
  border-left: 11px solid rgba(0,0,0,.4);
  -webkit-transition: all undefined ease-in-out;
  -moz-transition: all undefined ease-in-out;
}

<glamorous(span)
  arrow="top"
  bg={true}
  fg={false}
  position="left"
>
  <span
    className="glamor-0"
  />
</glamorous(span)>
`;

exports[`Arrow should render bg Arrow with position right and arrow bottom 1`] = `
.glamor-0,
[data-glamor-0] {
  position: absolute;
  content: "";
  transition: all undefined ease-in-out;
  z-index: 55;
  border-top: 9px solid transparent;
  border-bottom: 9px solid transparent;
  left: -11px;
  border-right: 11px solid rgba(0,0,0,.4);
  bottom: 0;
  -webkit-transition: all undefined ease-in-out;
  -moz-transition: all undefined ease-in-out;
}

<glamorous(span)
  arrow="bottom"
  bg={true}
  fg={false}
  position="right"
>
  <span
    className="glamor-0"
  />
</glamorous(span)>
`;

exports[`Arrow should render bg Arrow with position right and arrow center 1`] = `
.glamor-0,
[data-glamor-0] {
  position: absolute;
  content: "";
  transition: all undefined ease-in-out;
  z-index: 55;
  border-top: 9px solid transparent;
  border-bottom: 9px solid transparent;
  top: calc(50% - 9px);
  left: -11px;
  border-right: 11px solid rgba(0,0,0,.4);
  -webkit-transition: all undefined ease-in-out;
  -moz-transition: all undefined ease-in-out;
}

<glamorous(span)
  arrow="center"
  bg={true}
  fg={false}
  position="right"
>
  <span
    className="glamor-0"
  />
</glamorous(span)>
`;

exports[`Arrow should render bg Arrow with position right and arrow top 1`] = `
.glamor-0,
[data-glamor-0] {
  position: absolute;
  content: "";
  transition: all undefined ease-in-out;
  z-index: 55;
  border-top: 9px solid transparent;
  border-bottom: 9px solid transparent;
  top: 1px;
  left: -11px;
  border-right: 11px solid rgba(0,0,0,.4);
  -webkit-transition: all undefined ease-in-out;
  -moz-transition: all undefined ease-in-out;
}

<glamorous(span)
  arrow="top"
  bg={true}
  fg={false}
  position="right"
>
  <span
    className="glamor-0"
  />
</glamorous(span)>
`;

exports[`Arrow should render bg Arrow with position top and arrow center 1`] = `
.glamor-0,
[data-glamor-0] {
  position: absolute;
  content: "";
  transition: all undefined ease-in-out;
  z-index: 55;
  left: 50%;
  margin-left: -9px;
  border-left: 9px solid transparent;
  border-right: 9px solid transparent;
  bottom: -11px;
  border-top: 11px solid rgba(0,0,0,.4);
  -webkit-transition: all undefined ease-in-out;
  -moz-transition: all undefined ease-in-out;
}

<glamorous(span)
  arrow="center"
  bg={true}
  fg={false}
  position="top"
>
  <span
    className="glamor-0"
  />
</glamorous(span)>
`;

exports[`Arrow should render bg Arrow with position top and arrow left 1`] = `
.glamor-0,
[data-glamor-0] {
  position: absolute;
  content: "";
  transition: all undefined ease-in-out;
  z-index: 55;
  left: -1px;
  border-left: 9px solid transparent;
  border-right: 9px solid transparent;
  bottom: -11px;
  border-top: 11px solid rgba(0,0,0,.4);
  -webkit-transition: all undefined ease-in-out;
  -moz-transition: all undefined ease-in-out;
}

<glamorous(span)
  arrow="left"
  bg={true}
  fg={false}
  position="top"
>
  <span
    className="glamor-0"
  />
</glamorous(span)>
`;

exports[`Arrow should render bg Arrow with position top and arrow right 1`] = `
.glamor-0,
[data-glamor-0] {
  position: absolute;
  content: "";
  transition: all undefined ease-in-out;
  z-index: 55;
  border-left: 9px solid transparent;
  border-right: 9px solid transparent;
  bottom: -11px;
  border-top: 11px solid rgba(0,0,0,.4);
  right: -1px;
  -webkit-transition: all undefined ease-in-out;
  -moz-transition: all undefined ease-in-out;
}

<glamorous(span)
  arrow="right"
  bg={true}
  fg={false}
  position="top"
>
  <span
    className="glamor-0"
  />
</glamorous(span)>
`;

exports[`Arrow should render fg Arrow with default props 1`] = `
.glamor-0,
[data-glamor-0] {
  position: absolute;
  content: "";
  transition: all undefined ease-in-out;
  z-index: 60;
  -webkit-transition: all undefined ease-in-out;
  -moz-transition: all undefined ease-in-out;
}

<glamorous(span)
  fg={true}
>
  <span
    className="glamor-0"
  />
</glamorous(span)>
`;

exports[`Arrow should render fg Arrow with position bottom and arrow center 1`] = `
.glamor-0,
[data-glamor-0] {
  position: absolute;
  content: "";
  transition: all undefined ease-in-out;
  z-index: 60;
  left: 50%;
  margin-left: -8px;
  border-left: 8px solid transparent;
  border-right: 8px solid transparent;
  top: -10px;
  border-bottom: 10px solid #fff;
  -webkit-transition: all undefined ease-in-out;
  -moz-transition: all undefined ease-in-out;
}

<glamorous(span)
  arrow="center"
  bg={false}
  fg={true}
  position="bottom"
>
  <span
    className="glamor-0"
  />
</glamorous(span)>
`;

exports[`Arrow should render fg Arrow with position bottom and arrow left 1`] = `
.glamor-0,
[data-glamor-0] {
  position: absolute;
  content: "";
  transition: all undefined ease-in-out;
  z-index: 60;
  border-left: 8px solid transparent;
  border-right: 8px solid transparent;
  top: -10px;
  border-bottom: 10px solid #fff;
  -webkit-transition: all undefined ease-in-out;
  -moz-transition: all undefined ease-in-out;
}

<glamorous(span)
  arrow="left"
  bg={false}
  fg={true}
  position="bottom"
>
  <span
    className="glamor-0"
  />
</glamorous(span)>
`;

exports[`Arrow should render fg Arrow with position bottom and arrow right 1`] = `
.glamor-0,
[data-glamor-0] {
  position: absolute;
  content: "";
  transition: all undefined ease-in-out;
  z-index: 60;
  border-left: 8px solid transparent;
  border-right: 8px solid transparent;
  top: -10px;
  border-bottom: 10px solid #fff;
  right: 0px;
  -webkit-transition: all undefined ease-in-out;
  -moz-transition: all undefined ease-in-out;
}

<glamorous(span)
  arrow="right"
  bg={false}
  fg={true}
  position="bottom"
>
  <span
    className="glamor-0"
  />
</glamorous(span)>
`;

exports[`Arrow should render fg Arrow with position left and arrow bottom 1`] = `
.glamor-0,
[data-glamor-0] {
  position: absolute;
  content: "";
  transition: all undefined ease-in-out;
  z-index: 60;
  border-top: 8px solid transparent;
  border-bottom: 8px solid transparent;
  right: -10px;
  border-left: 10px solid #fff;
  bottom: 1px;
  -webkit-transition: all undefined ease-in-out;
  -moz-transition: all undefined ease-in-out;
}

<glamorous(span)
  arrow="bottom"
  bg={false}
  fg={true}
  position="left"
>
  <span
    className="glamor-0"
  />
</glamorous(span)>
`;

exports[`Arrow should render fg Arrow with position left and arrow center 1`] = `
.glamor-0,
[data-glamor-0] {
  position: absolute;
  content: "";
  transition: all undefined ease-in-out;
  z-index: 60;
  border-top: 8px solid transparent;
  border-bottom: 8px solid transparent;
  top: calc(50% - 8px);
  right: -10px;
  border-left: 10px solid #fff;
  -webkit-transition: all undefined ease-in-out;
  -moz-transition: all undefined ease-in-out;
}

<glamorous(span)
  arrow="center"
  bg={false}
  fg={true}
  position="left"
>
  <span
    className="glamor-0"
  />
</glamorous(span)>
`;

exports[`Arrow should render fg Arrow with position left and arrow top 1`] = `
.glamor-0,
[data-glamor-0] {
  position: absolute;
  content: "";
  transition: all undefined ease-in-out;
  z-index: 60;
  border-top: 8px solid transparent;
  border-bottom: 8px solid transparent;
  top: 2px;
  right: -10px;
  border-left: 10px solid #fff;
  -webkit-transition: all undefined ease-in-out;
  -moz-transition: all undefined ease-in-out;
}

<glamorous(span)
  arrow="top"
  bg={false}
  fg={true}
  position="left"
>
  <span
    className="glamor-0"
  />
</glamorous(span)>
`;

exports[`Arrow should render fg Arrow with position right and arrow bottom 1`] = `
.glamor-0,
[data-glamor-0] {
  position: absolute;
  content: "";
  transition: all undefined ease-in-out;
  z-index: 60;
  border-top: 8px solid transparent;
  border-bottom: 8px solid transparent;
  left: -10px;
  border-right: 10px solid #fff;
  bottom: 1px;
  -webkit-transition: all undefined ease-in-out;
  -moz-transition: all undefined ease-in-out;
}

<glamorous(span)
  arrow="bottom"
  bg={false}
  fg={true}
  position="right"
>
  <span
    className="glamor-0"
  />
</glamorous(span)>
`;

exports[`Arrow should render fg Arrow with position right and arrow center 1`] = `
.glamor-0,
[data-glamor-0] {
  position: absolute;
  content: "";
  transition: all undefined ease-in-out;
  z-index: 60;
  border-top: 8px solid transparent;
  border-bottom: 8px solid transparent;
  top: calc(50% - 8px);
  left: -10px;
  border-right: 10px solid #fff;
  -webkit-transition: all undefined ease-in-out;
  -moz-transition: all undefined ease-in-out;
}

<glamorous(span)
  arrow="center"
  bg={false}
  fg={true}
  position="right"
>
  <span
    className="glamor-0"
  />
</glamorous(span)>
`;

exports[`Arrow should render fg Arrow with position right and arrow top 1`] = `
.glamor-0,
[data-glamor-0] {
  position: absolute;
  content: "";
  transition: all undefined ease-in-out;
  z-index: 60;
  border-top: 8px solid transparent;
  border-bottom: 8px solid transparent;
  top: 2px;
  left: -10px;
  border-right: 10px solid #fff;
  -webkit-transition: all undefined ease-in-out;
  -moz-transition: all undefined ease-in-out;
}

<glamorous(span)
  arrow="top"
  bg={false}
  fg={true}
  position="right"
>
  <span
    className="glamor-0"
  />
</glamorous(span)>
`;

exports[`Arrow should render fg Arrow with position top and arrow center 1`] = `
.glamor-0,
[data-glamor-0] {
  position: absolute;
  content: "";
  transition: all undefined ease-in-out;
  z-index: 60;
  left: 50%;
  margin-left: -8px;
  border-left: 8px solid transparent;
  border-right: 8px solid transparent;
  bottom: -10px;
  border-top: 10px solid #fff;
  -webkit-transition: all undefined ease-in-out;
  -moz-transition: all undefined ease-in-out;
}

<glamorous(span)
  arrow="center"
  bg={false}
  fg={true}
  position="top"
>
  <span
    className="glamor-0"
  />
</glamorous(span)>
`;

exports[`Arrow should render fg Arrow with position top and arrow left 1`] = `
.glamor-0,
[data-glamor-0] {
  position: absolute;
  content: "";
  transition: all undefined ease-in-out;
  z-index: 60;
  border-left: 8px solid transparent;
  border-right: 8px solid transparent;
  bottom: -10px;
  border-top: 10px solid #fff;
  -webkit-transition: all undefined ease-in-out;
  -moz-transition: all undefined ease-in-out;
}

<glamorous(span)
  arrow="left"
  bg={false}
  fg={true}
  position="top"
>
  <span
    className="glamor-0"
  />
</glamorous(span)>
`;

exports[`Arrow should render fg Arrow with position top and arrow right 1`] = `
.glamor-0,
[data-glamor-0] {
  position: absolute;
  content: "";
  transition: all undefined ease-in-out;
  z-index: 60;
  border-left: 8px solid transparent;
  border-right: 8px solid transparent;
  bottom: -10px;
  border-top: 10px solid #fff;
  right: 0px;
  -webkit-transition: all undefined ease-in-out;
  -moz-transition: all undefined ease-in-out;
}

<glamorous(span)
  arrow="right"
  bg={false}
  fg={true}
  position="top"
>
  <span
    className="glamor-0"
  />
</glamorous(span)>
`;
